qutebrowser/browser/webkit/network/pac.py

```diff
@@ -0,0 +1,608 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# Copyright 2016 Florian Bruhin (The Compiler) <mail@qutebrowser.org>
+#
+# This file is part of qutebrowser.
+#
+# qutebrowser is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Evaluation of PAC scripts."""
+
+import sys
+import ipaddress
+import fnmatch
+import itertools
+import functools
+
+from PyQt5.QtCore import (QObject, QVariant, QMetaType, QDateTime,
+                          QEventLoop, pyqtSlot)
+from PyQt5.QtNetwork import (QNetworkProxy, QNetworkRequest, QHostInfo,
+                             QNetworkReply, QNetworkAccessManager)
+from PyQt5.QtQml import QJSEngine, QJSValue
+
+from qutebrowser.utils import log
+
+
+class ParseProxyError(Exception):
+
+    """Error in parsing PAC result string."""
+
+    pass
+
+
+class EvalProxyError(Exception):
+
+    """Error in evaluating PAC script."""
+
+    pass
+
+
+def _js_slot(*args):
+    """Wrap a methods as a JavaScript function.
+
+    Register a PACContext method as a JavaScript function, and catch
+    exceptions returning them as JavaScript Error objects.
+
+    Args:
+        args: Types of method arguments.
+
+    Return: Wrapped method.
+    """
+    def _decorator(method):
+        @functools.wraps(method)
+        def new_method(self, *args, **kwargs):
+            try:
+                return method(self, *args, **kwargs)
+            except:
+                e = str(sys.exc_info()[0])
+                # Supress logging for tests
+                #pylint: disable=protected-access
+                if self._fixed_dt is None:
+                    log.network.error("PAC evaluation error: {}".format(e))
```

Comment:
Would it make sense to use `log.network.exception` here so the stacktrace gets logged?